src/pudl_archiver/archivers/eia/eiasteo.py

```diff
@@ -0,0 +1,146 @@
+"""Archive EIA Short-Term Energy Outlook (STEO)."""
+
+import re
+from pathlib import Path
+from urllib.parse import urljoin, urlparse
+
+import pandas as pd
+
+from pudl_archiver.archivers.classes import (
+    AbstractDatasetArchiver,
+    ArchiveAwaitable,
+    ResourceInfo,
+)
+from pudl_archiver.frictionless import ZipLayout
+
+BASE_URL = "https://www.eia.gov/outlooks/steo/"
+
+ARCHIVES_URL = "https://www.eia.gov/outlooks/steo/outlook.php"
+CURRENT_URL = "https://www.eia.gov/outlooks/steo/data.php"
+
+# Prior to 1997 the STEO was quarterly.
+STEO_ARCHIVE_DATES = set(
+    [
+        str(q).lower()
+        for q in pd.period_range(start="1983-01", end="1996-12", freq="Q").asfreq(
+            "M", how="S"
+        )
+    ]
+    +
+    # The EIA STEO is released monthly.
+    [
+        str(q).lower()
+        for q in pd.period_range(start="1997-01", end=pd.to_datetime("today"), freq="M")
+    ]
+)
+
+
+class EiaSteoArchiver(AbstractDatasetArchiver):
+    """EIA STEO archiver."""
+
+    name = "eiasteo"
+
+    async def get_resources(self) -> ArchiveAwaitable:
+        """Download EIA STEO resources."""
+        # We grab data from the archive URL and the current URL separately.
+        # For each month, make a zipfile containing all relevant data from either the
+        # current or the archival URL
+
+        archival_soup = await self.get_soup(ARCHIVES_URL)
+        current_soup = await self.get_soup(CURRENT_URL)
+
+        archival_links = [
+            link
+            for link in archival_soup.find_all("a", href=True)
+            if (".pdf" in link["href"] or ".xls" in link["href"])
+            and ("archives" in link["href"])
+        ]
+        # Create a dictionary containing year_month partition and link href.
+        # All the links with "Q" are formatted 1Q89.pdf, e.g..
+        # to_datetime infers these as being 2089-Q1, so we take the year and the month
+        # from datetime and manually append it to the 1900 prefix.
+        # All the other links can be parsed to map sep09_filename.xls and sep09.pdf
+        # into 2009-09 partitions. We slice the first 5 characters to drop additional
+        # file name information.
+        # We use a dictionary here because some hrefs are duplicated, and should only
+        # have one partition per href.
+        archival_links = {
+            link["href"]: pd.to_datetime(
+                link.getText().split(".")[0][0:5], format="%b%y"
+            ).strftime("%Y-%m")
+            if "Q" not in link["href"]
+            else "19" + pd.to_datetime(link.getText().split(".")[0]).strftime("%y-%m")
```

Comment:
this is pretty dense string math! it's explained well in the comment but i wonder if a regex would be easier to read and explain.